internal/types/identical.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"bytes"
+	"errors"
+	"time"
+
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+)
+
+// Identical returns true if a and b are the same type
+// and has the same value.
+func Identical(a, b any) bool {
+	switch a := a.(type) {
+	case *Document:
+		b, ok := b.(*Document)
+		if !ok {
+			return false
+		}
+
+		if a.Len() != b.Len() {
+			return false
+		}
+
+		aIter, bIter := a.Iterator(), b.Iterator()
+		defer aIter.Close()
+		defer bIter.Close()
+
+		for {
+			_, aField, err := aIter.Next()
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				return true
+			} else if err != nil {
+				return false
```

Comment:
As this function doesn't have an error return value, I suggest we should panic in all those cases if `iter.Next` returns an error (for objects and arrays). Currently, we just ignore it, and this could cause unexpected behavior if such an error becomes possible in the future.

Reply:
Sounds sensible :hugs: let me do that